lib/metadata/linux/MiqRpmPackages.rb

```diff
@@ -1,152 +1,29 @@
+require_relative "MiqRpmPackages/Bdb"
+require_relative "MiqRpmPackages/Sqlite"
 
-  ENTRY_INFO = BinaryStruct.new([
-    'N',        'tag',
-    'N',        'ttype',
-    'N',        'offset',
-    'N',        'count'
-  ])
-  ENTRY_INFO_LEN = ENTRY_INFO.size
+class MiqRpmPackages
+  class << self
+    private
 
-  def initialize(fs, dbFile)
-    @pkgDb = MiqBerkeleyDB::MiqBdb.new(dbFile, fs)
-    # Pre-read all pages into the bdb cache, as we will be processing all of them anyway.
-    @pkgDb.readAllPages
+    alias orig_new new
   end
 
-  def each
-    @pkgDb.each_value do |v|
-      next if v.length <= RPML_HEADER_LEN
-
-      hdr = RPML_HEADER.decode(v)
-
-      offset = RPML_HEADER_LEN + (ENTRY_INFO_LEN * hdr['num_index'])
-      if v.length != offset + hdr['num_data']
-        $log.debug "record length = #{v.length}"
-        $log.debug "num_index = #{hdr['num_index']}"
-        $log.debug "num_data = #{hdr['num_data']}"
-        $log.error "Invalid or corrupt RPM database record"
-        next
-      end
-
-      data = v[offset, hdr['num_data']]
-      pkg = {}
-
-      eis = ENTRY_INFO.decode(v[RPML_HEADER_LEN..-1], hdr['num_index'])
-      eis.each do |ei|
-        tag = TAGIDS[ei['tag']]
-        next if tag.nil?
-        pkg[tag] = getVal(data, ei)
-        pkg[tag] = convert(tag, pkg[tag])
+  def self.new(fs, dbDir)
+    if self == MiqRpmPackages
+      if fs.fileExists?(File.join(dbDir, "Packages"))
+        MiqRpmPackages::Bdb.new(fs, File.join(dbDir, "Packages"))
+      elsif fs.fileExists?(File.join(dbDir, "rpmdb.sqlite"))
+        MiqRpmPackages::Sqlite.new(fs, File.join(dbDir, "rpmdb.sqlite"))
+      else
+        raise ArgumentError, "Invalid RPM database"
       end
-      pkg['installed'] = true unless pkg.empty?
-      yield(MiqHashStruct.new(pkg))
-    end
-  end # def each
-
-  def close
-    @pkgDb.close
-  end
-
-  private
-
-  def time_tag?(tag)
-    ['installtime', 'buildtime'].include?(tag)
-  end
-
-  def convert(tag, val)
-    time_tag?(tag) ? Time.at(val).utc : val
-  end
-
-  def getVal(data, ei)
-    case ei['ttype']
-    when RPM_INT32_TYPE         then return(getInt32Val(data, ei['offset']))
-    when RPM_STRING_TYPE        then return(getStringVal(data, ei['offset']))
-    when RPM_STRING_ARRAY_TYPE  then return(getStringArray(data, ei['offset'], ei['count']).join("\n"))
-    when RPM_I18NSTRING_TYPE    then return(getStringArray(data, ei['offset'], ei['count']).join("\n").AsciiToUtf8)
     else
-      $log.warn "MiqRpmPackages.getVal: unsupported data type: #{ei['ttype']}"
-      return("")
-    end
-  end
-
-  def getInt32Val(data, offset)
-    (data[offset, 4].unpack("N").first)
-  end
-
-  def getStringVal(data, offset)
-    eos = data.index(0.chr, offset) - offset
-    (data[offset, eos])
-  end
-
-  def getStringArray(data, offset, count)
-    ra = []
-    cpos = offset
-
-    count.times do
-      s = getStringVal(data, cpos)
-      ra << s
-      cpos += (s.length + 1)
+      orig_new(fs, dbDir)
```

Comment:
I think we can use a `super` pattern here instead of the orig_new

Reply:
Even so, I will merge without this.

Reply:
I pulled this pattern from linux_admin, if this isn't the best way to do it we should probably change both then